Modern logging without breaking change

fredericDelaporte opened the pull request:
Thinking about it, finally I consider the more straightforward non-breaking way is to have the new interface inheriting the old one, fully obsoleted. It even avoids the dummy extension methods meant to catch calls with exception last: they now fall on the obsolete method.

fredericDelaporte committed:
Modern logging without breaking changes.

fredericDelaporte commented on src/NHibernate.Test/Logging/LoggerPerformanceTest.cs:
```diff
@@ -109,24 +109,26 @@ public void WarnFormat(string format, params object[] args)
 		private class MockNHibernateLoggerFactory : INHibernateLoggerFactory
 		{
-			public IInternalLogger2 LoggerFor(string keyName)
+			public INHibernateLogger LoggerFor(string keyName)
```
Having seen the factory name, I found it nicer than the 2 suffix and choose a similar naming for the logger.

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -8,36 +8,90 @@
-	public interface IInternalLogger2
+	// Required till look alike methods taking object are dropped.
+	[Obsolete("Implement and use NHibernate.INHibernateLogger")]
+	public interface ITransitionalInternaLogger : IInternalLogger
```
Ok, I had to add some other dummy methods (but less dummy: they work) with this solution. Since we obsoleting passing an object as a message, and that the instance methods takes precedence other extensions even if extensions have a better matching signature, we need these methods for being able to call something not obsoleted in this case and without using code tricks.

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -85,16 +159,16 @@ private static INHibernateLoggerFactory GetLoggerFactory(string nhibernateLogger
-#pragma warning disable 618
-				if (loadedLoggerFactory is ILoggerFactory oldStyleFactory)
+				var loadedLoggerFactory = Activator.CreateInstance(loggerFactoryType);
+				if (loadedLoggerFactory is INHibernateLoggerFactory newStyleFactory)
```
Test inverted because log4net logger factory does now implement both interfaces (old and new).

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -53,29 +107,49 @@ public interface IInternalLogger2
+	/// By default, it will use a <see cref="Log4NetLoggerFactory"/> if log4net is available, otherwise it will
+	/// use a <see cref="NoLoggingLoggerFactory"/>.
+	/// </remarks>
+	public static class LoggerProvider
```
Switched to static: constructor was private (so not a breaking change) and it simplifies the implementation.

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -272,36 +358,158 @@ public bool IsEnabled(InternalLogLevel logLevel)
+	/// </summary>
+	// Since 5.1
+	[Obsolete("To be used as logger base class for implementing the obsolete interface ITransitionalInternaLogger until it gets dropped.")]
+	public abstract class NHibernateLoggerBase
```
This is a base class meant to be drop once we drop the old logger. A bit special way of handling obsolete, but this allows implementors of the new interface to only implement its two new methods. The base class implements all the obsolete ones (and the five "semi-dummy" ones). When it will be dropped, implementors will just need to cease inheriting the base class.

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -272,36 +358,158 @@ public bool IsEnabled(InternalLogLevel logLevel)
+	/// <summary>
+	/// Factory building loggers which log nothing.
+	/// </summary>
 	public class NoLoggingLoggerFactory: INHibernateLoggerFactory
```
Still a breaking change, I am going to add another commit. It must still implement the old interface too.

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -272,36 +358,158 @@ public bool IsEnabled(InternalLogLevel logLevel)
+	/// <summary>
+	/// Logger which logs nothing.
+	/// </summary>
+	public class NoLoggingInternalLogger: NHibernateLoggerBase, INHibernateLogger
```
Adding a base class or changing an interface for a more derived one is not a breaking change.

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -138,32 +212,43 @@ private static string GetNhibernateLoggerClass()
+		/// </summary>
+		/// <param name="keyName">The log key.</param>
+		/// <returns>A NHibernate logger.</returns>
+		public static INHibernateLogger LoggerFor(string keyName)
```
Yielding a more derived interface is not a breaking change.

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -315,17 +523,31 @@ static Log4NetLoggerFactory()
 		}
 #pragma warning restore 618
 
+		INHibernateLogger INHibernateLoggerFactory.LoggerFor(string keyName)
```
Once the old interface dropped, we could turn it to implicit implementation. (Same for next.)

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -404,232 +629,244 @@ static Log4NetLogger()
+		#region IInternalLogger re-implementations required for avoiding a breaking change
+		// If some dependent library use the old internal logger, passing to it objects while using
+		// formatters, we must then implement directly the call. The base impl would instead call
+		// ToString on the object, preventing the formatters to work.
```
Special case here till the old interface is not dropped, for avoiding a subtle breaking change.

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -404,232 +629,244 @@ static Log4NetLogger()
 		{
-			if (IsFatalEnabled)
-				FatalDelegate(logger, message);
+			FatalDelegate(_logger, message);
```
Since these methods are obsolete anyway, I do not care about perf here (and log4net does the check itself moreover).

fredericDelaporte reviewed (commented):
@hazzik, I propose an alternate way for avoiding breaking changes for the modern logging. This a PR on ngbrown repository.

ngbrown commented:
@fredericDelaporte Thanks for doing this as a pull request.

Does an application implementing either the old logger or the new logger have to implement the new base class?

ngbrown commented:
I pushed all the obvious stuff from your commits. I'll need some time to figure out what you are trying to do with `NHibernateLoggerBase`.

fredericDelaporte commented:
> Does an application implementing either the old logger or the new logger have to implement the new base class?

No, it is just a helper for reducing the work of implementing a new logger until its obsolete interface ancestors are dropped.

The aim is to avoid breaking current `GetLogger` while allowing it to still yield the new interface.

fredericDelaporte commented:
I have rebased on your PR.

> > Does an application implementing either the old logger or the new logger have to implement the new base class?

> No, it is just a helper for reducing the work of implementing a new logger until its obsolete interface ancestors are dropped.
>
> The aim is to avoid breaking current `GetLogger` while allowing it to still yield the new interface.

So more details: in this PR, the new logger interface `INHibernateLogger` derives from the old one, `IInternalLogger`. This allows `GetLogger` to yield the new interface without this change being breaking, since it is still by the same time an `IInternalLogger`.

But this forces implementors to provides both old and new methods. For reducing the work of implementing old methods, I have added a base class providing default implementation, defaulting to calling the new interface methods. These implementations are certainly not the most performant runtime wise, but at they supports only obsolete methods no more used in NHibernate, I do not care.

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -76,7 +76,24 @@ public interface IInternalLogger
+	// Since v5.1
+	// Required till look alike methods taking object are dropped.
+	[Obsolete("Implement and use NHibernate.INHibernateLogger")]
+	public interface ITransitionalInternaLogger : IInternalLogger
```
Ok, I had to add some other dummy methods (but less dummy: they work) with this solution. Since we obsoleting passing an object as a message, and that the instance methods takes precedence other extensions even if extensions have a better matching signature, we need these methods for being able to call something not obsoleted in this case and without using code tricks.

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -126,11 +143,10 @@ public interface INHibernateLoggerFactory
 	/// use a <see cref="NoLoggingLoggerFactory"/>.
 	/// </remarks>
-	public class LoggerProvider
+	public static class LoggerProvider
```
Switched to static: constructor was private (so not a breaking change) and it simplifies the implementation.

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -146,15 +162,15 @@ private static INHibernateLoggerFactory GetLoggerFactory(string nhibernateLogger
 				var loadedLoggerFactory = Activator.CreateInstance(loggerFactoryType);
-#pragma warning disable 618
-				if (loadedLoggerFactory is ILoggerFactory oldStyleFactory)
+				if (loadedLoggerFactory is INHibernateLoggerFactory newStyleFactory)
```
Test inverted because log4net logger factory does now implement both interfaces (old and new).

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -76,7 +76,24 @@ public interface IInternalLogger
+	/// <summary>
+	/// NHibernate internal logger interface.
+	/// </summary>
+	public interface INHibernateLogger: ITransitionalInternaLogger
```
The new interface is still a `IInternalLogger` for avoiding a breaking change.
For allowing transitioning from obsolete methods to non obsolete ones, which are mostly extensions methods, I was forced for the "just a string" log case to introduce additional not yet obsolete methods, but which will be drop-able without breaking compilation (still a binary break of course) once the other obsoletes are dropped. Thus the intermediate interface `ITransitionalInternalLogger` already obsolete.

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -498,114 +656,139 @@ public Log4NetLogger(object logger)
 		{
-			if (IsFatalEnabled)
-				FatalDelegate(_logger, message);
+			FatalDelegate(_logger, message);
```
Since these methods are obsolete anyway, I do not care about perf here (and log4net does the check itself moreover).

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -349,39 +360,174 @@ public bool IsEnabled(InternalLogLevel logLevel)
+	/// </summary>
+	// Since 5.1
+	[Obsolete("To be used as logger base class for implementing the obsolete interface ITransitionalInternaLogger until it gets dropped.")]
+	public abstract class NHibernateLoggerBase
```
This is a base class meant to be drop once we drop the old logger. A bit special way of handling obsolete, but this allows implementors of the new interface to only implement its two new methods. The base class implements all the obsolete ones (and the five "semi-dummy" ones). When it will be dropped, implementors will just need to cease inheriting the base class.

This is just optional, using it is not mandatory.

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -349,39 +360,174 @@ public bool IsEnabled(InternalLogLevel logLevel)
+		protected NHibernateLoggerBase()
+		{
+			_this = this as INHibernateLogger ??
+				throw new InvalidOperationException("Concrete implementation must be an INHibernateLogger");
```
I have decided to not have this base class directly implementing `INHibernateLogger` in order to avoid having to supply its two new methods in the base class as abstract, which would force them to be virtual.

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -349,39 +360,174 @@ public bool IsEnabled(InternalLogLevel logLevel)
+	/// <summary>
+	/// Logger which logs nothing.
+	/// </summary>
+	public class NoLoggingInternalLogger: NHibernateLoggerBase, INHibernateLogger
```
Adding a base class or changing an interface for a more derived one is not a breaking change.

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -398,16 +544,28 @@ static Log4NetLoggerFactory()
 		}
 #pragma warning restore 618
 
+		INHibernateLogger INHibernateLoggerFactory.LoggerFor(string keyName)
```
Once the old interface dropped, we could turn it to implicit implementation. (Same for next.)

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -498,114 +656,139 @@ public Log4NetLogger(object logger)
+		#region IInternalLogger re-implementations required for avoiding a breaking change
+		// If some dependent library use the old internal logger, passing to it objects while using
+		// formatters, we must then implement directly the call. The base impl would instead call
+		// ToString on the object, preventing the formatters to work.
```
Special case here till the old interface is not dropped, for avoiding a subtle breaking change.

fredericDelaporte commented on src/NHibernate.Test/Logging/LoggerPerformanceTest.cs:
```diff
@@ -120,13 +120,15 @@ public INHibernateLogger LoggerFor(System.Type type)
-			private bool IsDebugEnabled { get; } = false;
-			private bool IsInfoEnabled { get; } = false;
-			private bool IsWarnEnabled { get; } = true;
+			private new bool IsErrorEnabled { get; } = true;
```
That hides the base class member, but we do not care. The base class member still serves as the interface member.

fredericDelaporte commented on src/NHibernate/Logging.cs:
```diff
@@ -250,12 +261,12 @@ public LegacyLoggerFactoryAdaptor(ILoggerFactory factory)
 			}
 
-			public INHibernateLogger LoggerFor(string keyName)
+			INHibernateLogger INHibernateLoggerFactory.LoggerFor(string keyName)
```
Just avoids a R# warning about hiding the `LoggerFor` of the encapsulating class.

fredericDelaporte reviewed (commented):
Adding some more comments and reporting comments from previous review.

fredericDelaporte committed:
Some more explanations in code comments.

ngbrown commented:
> the new logger interface `INHibernateLogger` derives from the old one, `IInternalLogger`....But this forces implementors to provides both old and new methods.

So when `IInternalLogger` is removed from the a future version of NHibernate, any implementations will be forced to update at that point as it would be a hard breaking change?

I was thinking there would be a way to do this that if the implementors didn't have any warnings in v5.1, it would continue to work fine in v6.0.

fredericDelaporte commented:
> So when IInternalLogger is removed from a future version of NHibernate, any implementations will be forced to update at that point as it would be a hard breaking change?

Not forcibly, we could keep in place the base class but emptied. So in fact I should not obsolete it right now, for avoiding having implementors disabling the warning. It would be obsoleted when old interfaces will be removed. Going to change that.

> I was thinking there would be a way to do this that if the implementors didn't have any warnings in v5.1, it would continue to work fine in v6.0.

With current setup they have a warning, since the base class is already obsolete. But better remove it and do it two steps: when removing the old interfaces, keep the base class but then obsolete it (and empty it since it would contain only obsolete methods or methods having a fallback on extension methods).

fredericDelaporte committed:
Adjustment to obsolete handling.
